### PR TITLE
Trie decoupling

### PR DESCRIPTION
**Pull request by colltoaction**
* Decouples getting a snapshot from the trie
* Decouples saving from the trie

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/db/RepositoryLocator.java [hunk 1]
```diff
@@ -45,6 +59,11 @@ public Repository startTrackingAt(BlockHeader header) {
 
     private MutableTrie mutableTrieSnapshotAt(BlockHeader header) {
         Keccak256 stateRoot = stateRootHandler.translate(header);
-        return new MutableTrieImpl(repository.getTrie().getSnapshotTo(stateRoot));
+
+        if (EMPTY_HASH.equals(stateRoot)) {
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/db/RepositoryLocator.java]: I cannot find this behavior in the `co.rsk.db.StateRootHandler#translate` or in the context of this commit, how did you get this?

File: rskj-core/src/main/java/co/rsk/db/migration/OrchidToUnitrieMigrator.java [hunk 2]
```diff
@@ -232,19 +232,13 @@ private void buildPartialUnitrie(Trie orchidAccountsTrie, Repository repository)
         }
     }
 
-    private void migrateContract(RskAddress accountAddress, Repository currentRepository, byte[] contractDataRaw, byte[] accountCodeHash, byte[] stateRoot) throws MissingContractStorageKeysException {
+    private void migrateContract(RskAddress accountAddress, Repository currentRepository, byte[] contractDataRaw, byte[] accountCodeHash, byte[] oldAccountStateRoot) throws MissingContractStorageKeysException {
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/db/migration/OrchidToUnitrieMigrator.java]: I think we can use `orchidAccountStateRoot` here

File: rskj-core/src/main/java/co/rsk/db/migration/OrchidToUnitrieMigrator.java [hunk 3]
```diff
@@ -303,6 +297,24 @@ private void migrateContract(RskAddress accountAddress, Repository currentReposi
         }
     }
 
+    private Trie getContractStorageTrie(RskAddress contractAddress, ContractData contractData, Keccak256 oldAccountStateRoot) {
+        try {
+            Trie trie = getContractStorageTrie(contractAddress, contractData);
+            TrieStore trieStore = trie.getStore();
+            if (Arrays.equals(EMPTY_TRIE_HASH, oldAccountStateRoot.getBytes())) {
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/db/migration/OrchidToUnitrieMigrator.java]: I think this has been checked in `co.rsk.db.migration.OrchidToUnitrieMigrator#migrateContract`, isn't it?

File: rskj-core/src/main/java/co/rsk/db/migration/OrchidToUnitrieMigrator.java [hunk 4]
```diff
@@ -303,6 +297,24 @@ private void migrateContract(RskAddress accountAddress, Repository currentReposi
...
+                return new Trie(trieStore);
+            }
+
+            if (trie.isSaved() && trie.getHash().equals(oldAccountStateRoot)) {
+                return trie;
+            }
+
+            return trieStore.retrieve(oldAccountStateRoot.getBytes());
+        } catch (IllegalArgumentException e) {
+            throw new IllegalStateException("Cannot find state root trie. Check the log for more info.", e);
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/db/migration/OrchidToUnitrieMigrator.java]: Haha, I just realized this nonsense exception message.
> Check the log for more info

🤔

File: rskj-core/src/main/java/co/rsk/trie/TrieStoreImpl.java [hunk 5]
```diff
@@ -44,10 +44,29 @@ public TrieStoreImpl(KeyValueDataSource store) {
      */
     @Override
     public void save(Trie trie) {
-        this.store.put(trie.getHash().getBytes(), trie.toMessage());
+        save(trie, true);
+    }
+
+    private void save(Trie trie, boolean forceSaveRoot) {
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/trie/TrieStoreImpl.java]: I think we need a nice comment here about the meaning of `forceSaveRoot` and why we need it

File: rskj-core/src/main/java/co/rsk/trie/TrieStoreImpl.java [hunk 6]
```diff
@@ -44,10 +44,29 @@ public TrieStoreImpl(KeyValueDataSource store) {
      */
     @Override
     public void save(Trie trie) {
-        this.store.put(trie.getHash().getBytes(), trie.toMessage());
+        save(trie, true);
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/trie/TrieStoreImpl.java]: I think this hunk belongs to the next commit.

File: rskj-core/src/main/java/co/rsk/trie/TrieStore.java [hunk 7]
```diff
@@ -21,8 +21,6 @@
 public interface TrieStore {
     void save(Trie trie);
 
-    void saveValue(Trie trie);
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/trie/TrieStore.java]: Nice!

File: rskj-core/src/main/java/co/rsk/trie/TrieStoreImpl.java [hunk 8]
```diff
@@ -31,24 +35,22 @@
  */
 public class TrieStoreImpl implements TrieStore {
 
-    // a key value data source to use
     private KeyValueDataSource store;
 
+    /** Weak references are removed once the tries are garbage collected */
+    private Set<Trie> savedTries = Collections.newSetFromMap(new WeakHashMap<>());
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/trie/TrieStoreImpl.java]: You deserve this
dca

File: rskj-core/src/main/java/co/rsk/db/RepositoryLocator.java (hunk 1 above)
Comment (collheesi) [reply rskj-core/src/main/java/co/rsk/db/RepositoryLocator.java]: From here:

```java
    public Trie getSnapshotTo(Keccak256 hash) {
        if (emptyHash.equals(hash)) {
            return new Trie(this.store);
        }
```

File: rskj-core/src/main/java/co/rsk/trie/TrieStoreImpl.java (hunk 8 above)
Comment (collheesi) [reply rskj-core/src/main/java/co/rsk/trie/TrieStoreImpl.java]: 😍

File: rskj-core/src/main/java/co/rsk/db/migration/OrchidToUnitrieMigrator.java [hunk 9]
```diff
@@ -352,13 +345,11 @@ private static Trie orchidTrieDeserialize(byte[] bytes) {
         byte[] root = Arrays.copyOfRange(bytes, Short.BYTES, expectedSize);
         TrieStore store = orchidTrieStoreDeserialize(bytes, expectedSize, new HashMapDB());
 
-        Trie newTrie = store.retrieve(root);
-
-        if (newTrie == null) {
+        if (store.retrieve(root) == null) {
```
Comment (lsebrie) [inline rskj-core/src/main/java/co/rsk/db/migration/OrchidToUnitrieMigrator.java]: Check if this still can be null

**Pull request merged**